test/distributed/org/apache/cassandra/distributed/action/GossipHelper.java

```diff
@@ -0,0 +1,431 @@
+    }
+
+    public static VersionedApplicationState versionedToken(IInvokableInstance instance, ApplicationState applicationState, SerializableBiFunction<IPartitioner, Collection<Token>, VersionedValue> supplier)
+    {
+        return unsafeVersionedToken(instance, applicationState, supplier, instance.config().getString("partitioner"), instance.config().getString("initial_token"));
+    }
+
+    public void removeFromRing(IInstance peer)
+    {
+        InetAddressAndPort endpoint = toCassandraInetAddressAndPort(peer.broadcastAddress());
+
+        TokenMetadata tokenMetadata = StorageService.instance.getTokenMetadata();
+        tokenMetadata.removeEndpoint(endpoint);
+        Collection<Token> tokens = tokenMetadata.getTokens(endpoint);
+        if (!tokens.isEmpty())
+            tokenMetadata.removeBootstrapTokens(tokens);
+
+        Gossiper.runInGossipStageBlocking(() -> Gossiper.instance.unsafeAnulEndpoint(endpoint));
+
+        SystemKeyspace.removeEndpoint(endpoint);
+        PendingRangeCalculatorService.instance.update();
+        PendingRangeCalculatorService.instance.blockUntilFinished();
+    }
+
+    public static void changeGossipState(Stream<IInvokableInstance> targets, Instance peer, List<VersionedApplicationState> newState)
+    {
+        targets.forEach((instance) -> {
+            changeGossipState(instance, peer, newState);
+        });
+    }
+
+    /**
+     * Changes gossip state of the `peer` on `target`
+     */
+    public static void changeGossipState(IInvokableInstance target, IInstance peer, List<VersionedApplicationState> newState)
+    {
+        int version = peer.isShutdown() ? MessagingService.current_version : peer.getMessagingVersion();
+
+        target.appliesOnInstance((InetSocketAddress addr, UUID hostId, Integer messagingVersion) -> {
+            InetAddressAndPort endpoint = toCassandraInetAddressAndPort(addr);
+            StorageService storageService = StorageService.instance;
+
+            Gossiper.runInGossipStageBlocking(() -> {
+                EndpointState state = Gossiper.instance.getEndpointStateForEndpoint(endpoint);
+                if (state == null)
+                {
+                    Gossiper.instance.initializeNodeUnsafe(endpoint, hostId, 1);
+                    state = Gossiper.instance.getEndpointStateForEndpoint(endpoint);
+                    if (state.isAlive() && !Gossiper.instance.isDeadState(state))
+                        Gossiper.instance.realMarkAlive(endpoint, state);
+                }
+
+                for (VersionedApplicationState value : newState)
+                {
+                    ApplicationState as = toApplicationState(value);
+                    VersionedValue vv = toVersionedValue(value);
+                    state.addApplicationState(as, vv);
+                    storageService.onChange(endpoint, as, vv);
+                }
+            });
+
+            MessagingService.instance().versions.set(endpoint, messagingVersion);
```

Comment:
Can we remove this and rely on `NET_VERSION`?  We get this for free if we add `NET_VERSION`.

Reply:
Sure, makes sense. Switched and removed.